### PR TITLE
Add endpoint to update molecular mechanism

### PR DESCRIPTION
**Pull request by dglemos**
Ticket: [G2P-182](https://www.ebi.ac.uk/panda/jira/browse/G2P-182)
We only allow updates if the mechanism is 'undetermined'.

**New endpoint**
`gene2phenotype/api/lgd/<str:stable_id>/update_mechanism/`

The input data has the same format as the mechanism data sent by the curation page.

**To test** 
db: dlemos_G2P_2024_18
Input data:
```
{
    "molecular_mechanism": {
				"name": "gain of function",
				"support": "evidence"
		    },
    "mechanism_synopsis": {
				"name": "",
				"support": ""
		    },
    "mechanism_evidence": [
        {
						"pmid": "18247425",
						"description": "This is description for the mechanism evidence.",
						"evidence_types": [
								{
										"primary_type": "Function",
										"secondary_type": [
												"Biochemical"
										]
								},
								{
										"primary_type": "Rescue",
										"secondary_type": [
												"Human",
												"Patient Cells"
										]
								}
						]
				}
    ]
}
```

### CODE REVIEW

File: gene2phenotype_project/gene2phenotype_app/views/locus_genotype_disease.py [hunk 1]
```diff
@@ -218,6 +219,150 @@ def update(self, request, stable_id):
...
+        g2p_stable_id = get_object_or_404(G2PStableID, stable_id=stable_id)
+        # Get the record
+        queryset = LocusGenotypeDisease.objects.filter(stable_id=g2p_stable_id, is_deleted=0)
+
+        if not queryset.exists():
+            self.handle_no_permission('Entry', stable_id)
+        else:
+            # Only 'undetermined' mechanisms can be updated
+            lgd_obj = queryset.first()
+            if lgd_obj.molecular_mechanism.mechanism.value != "undetermined":
```
Comment (olaaustine) [inline gene2phenotype_project/gene2phenotype_app/views/locus_genotype_disease.py:253]: Suggesting that this is 
if lgd_obj.molecular_mechanism.mechanism.value != "undetermined" and lgd_obj.molecular_mechanism.mechanism_support.value != "inferred":

File: gene2phenotype_project/gene2phenotype_app/serializers/locus_genotype_disease.py [hunk 2]
```diff
@@ -447,6 +452,134 @@ def update(self, instance, validated_data):
...
+            is_deleted = 0
+        )
+
+        # Get evidence - the mechanism evidence was validated in the view 'LGDUpdateMechanism'
+        # Example: {'pmid': '25099252', 'description': 'text', 'evidence_types': 
+        #          [{'primary_type': 'Rescue', 'secondary_type': ['Human', 'Patient Cells']}]}
+        for evidence in mechanism_evidence:
+            pmid = evidence.get("pmid")
+
+            try:
```
Comment (olaaustine) [inline gene2phenotype_project/gene2phenotype_app/serializers/locus_genotype_disease.py:534]: i think we should add the publication if it does not exist

**Review by olaaustine — COMMENTED** [review]
Minor comments, we also already discussed most of them but just to note some for the future. 
Thanks @dglemos

File: gene2phenotype_project/gene2phenotype_app/serializers/locus_genotype_disease.py (hunk 2 above)
Comment (dglemos) [reply gene2phenotype_project/gene2phenotype_app/serializers/locus_genotype_disease.py:537]: Yes, I added a note to do that in the future.

File: gene2phenotype_project/gene2phenotype_app/serializers/locus_genotype_disease.py [hunk 3]
```diff
@@ -447,6 +452,138 @@ def update(self, instance, validated_data):
...
+                            "support": ""
+                        },
+                        "mechanism_evidence": [{'pmid': '25099252', 'description': 'text', 'evidence_types': 
+                            [{'primary_type': 'Rescue', 'secondary_type': ['Human', 'Patient Cells']}]}]
+
+        """
+        molecular_mechanism_value = validated_data.get("molecular_mechanism")["name"]
+        molecular_mechanism_support = validated_data.get("molecular_mechanism")["support"]
+        mechanism_synopsis = validated_data.get("mechanism_synopsis", None) # synopsis is optional
+        mechanism_evidence = validated_data.get("mechanism_evidence", None) # evidence is optional
```
Comment (seeta-ramaraju) [inline gene2phenotype_project/gene2phenotype_app/serializers/locus_genotype_disease.py:478]:
```suggestion
        mechanism_evidence = validated_data.get("mechanism_evidence", None) # synopsis evidence is optional
```

Comment (dglemos) [reply gene2phenotype_project/gene2phenotype_app/serializers/locus_genotype_disease.py:478]: `mechanism_evidence` is the molecular mechanism evidence. The synopsis has no evidence directly linked to it.

Comment (dglemos) [reply gene2phenotype_project/gene2phenotype_app/serializers/locus_genotype_disease.py:478]: I'm going to improve these comments.

File: gene2phenotype_project/gene2phenotype_app/urls.py [hunk 4]
```diff
@@ -39,6 +39,8 @@ def perform_create(self, serializer):
 
     ### Endpoints to update/add/delete the G2P record (LGD) ###
     path('lgd/<str:stable_id>/update_confidence/', views.LGDUpdateConfidence.as_view(), name="lgd_update_confidence"),
+    # Update molecular mechanism - only allows to update if mechanism is 'undetermined'
```
Comment (seeta-ramaraju) [inline gene2phenotype_project/gene2phenotype_app/urls.py:42]:
```suggestion
    # Update molecular mechanism - only allows to update if mechanism is 'undetermined' and evidence is 'inferred'
```

File: gene2phenotype_project/gene2phenotype_app/views/locus_genotype_disease.py (hunk 1 above)
Comment (seeta-ramaraju) [reply gene2phenotype_project/gene2phenotype_app/views/locus_genotype_disease.py:253]: I think to be clear it should be (Option A)
`if not (lgd_obj.molecular_mechanism.mechanism.value == "undetermined" and lgd_obj.molecular_mechanism.mechanism_support.value == "inferred"):`

The above expression is also equivalent to saying (Option B)
`if lgd_obj.molecular_mechanism.mechanism.value != "undetermined" or lgd_obj.molecular_mechanism.mechanism_support.value != "inferred":`

But Option A seems more clear that if it the mechanism and support combination is not ("undetermined" and "inferred") then we cannot update the mechanism

What are your thoughts?

File: gene2phenotype_project/gene2phenotype_app/views/locus_genotype_disease.py [hunk 5]
```diff
@@ -218,6 +219,149 @@ def update(self, request, stable_id):
...
+
+        # Separate method to update mechanism
+        # Updating the mechanism can be complex, specially if evidence data is provided
+        # To avoid problems with other LDG updates, the mechanism is going to be
+        # updated in a separate method - this implies extra validation
+        try:
+            serializer.update_mechanism(lgd_obj, mechanism_data)
+        except Exception as e:
+            return Response(
+                {"error": f"Problem updating molecular mechanism"},
```
Comment (seeta-ramaraju) [inline gene2phenotype_project/gene2phenotype_app/views/locus_genotype_disease.py:357]:
```suggestion
                {"error": f"Error while updating molecular mechanism"},
```

File: gene2phenotype_project/gene2phenotype_app/serializers/locus_genotype_disease.py [hunk 6]
```diff
@@ -447,6 +452,138 @@ def update(self, instance, validated_data):
 
         return instance
 
+    def update_mechanism(self, lgd_instance, validated_data):
+        """
+            Method to update the molecular mechanism of the LGD record.
+            It only allows to update mechanisms with value 'undetermined'
+            and evidence 'inferred'.
```
Comment (seeta-ramaraju) [inline gene2phenotype_project/gene2phenotype_app/serializers/locus_genotype_disease.py:459]:
```suggestion
            It only allows to update mechanisms with value 'undetermined'
            and support value 'inferred'.
```

Comment (seeta-ramaraju) [reply gene2phenotype_project/gene2phenotype_app/serializers/locus_genotype_disease.py:459]: just to be clear

**Review by seeta-ramaraju — APPROVED** [review]
LGTM

**Review by seeta-ramaraju — APPROVED** [review]
LGTM

**Pull request merged**